src/main/java/org/killbill/billing/plugin/helloworld/HelloWorldInvoicePluginApi.java

```diff
@@ -0,0 +1,151 @@
+			final Iterable<PluginProperty> properties, final CallContext callContext) {
+
+		UUID accountId = newInvoice.getAccountId();
+		Account account = getAccount(accountId, callContext);
+		Set<Invoice> allInvoices = getAllInvoicesOfAccount(account, newInvoice, callContext);
+		ImmutableList.Builder<InvoiceItem> additionalItems = ImmutableList.builder();
+
+		// Creating tax item for first Item of new Invoice
+		List<InvoiceItem> newInvoiceItems = newInvoice.getInvoiceItems();
+		InvoiceItem newInvoiceItem = newInvoiceItems.get(0);
+		BigDecimal charge = new BigDecimal("80");
+		InvoiceItem taxItem = PluginInvoiceItem.createTaxItem(newInvoiceItem, newInvoiceItem.getInvoiceId(),
+				newInvoice.getInvoiceDate(), null, charge, "Tax Item");
+		additionalItems.add(taxItem);
+
+		// Creating External Charge for first Item of new Invoice
+		InvoiceItem externalItem = PluginInvoiceItem.create(newInvoiceItem, newInvoiceItem.getInvoiceId(),
+				newInvoice.getInvoiceDate(), null, charge, "External Item", InvoiceItemType.EXTERNAL_CHARGE);
+		additionalItems.add(externalItem);
+
+		// Creating adjustment item for first Item of Historical Invoice
+		for (Invoice invoice : allInvoices) {
+			if (!invoice.getId().equals(newInvoice.getId())) {
+				List<InvoiceItem> invoiceItems = invoice.getInvoiceItems();
+				// Check for if any adjustment item exists for Historical Invoice
+				if (checkforAdjustmentItem(invoiceItems)) {
+					break;
+				}
+				for (InvoiceItem item : invoiceItems) {
+					if (isTaxItem(item)) {
+						charge = new BigDecimal("-40");
+						InvoiceItem adjItem = PluginInvoiceItem.createAdjustmentItem(item, item.getInvoiceId(),
+								newInvoice.getInvoiceDate(), newInvoice.getInvoiceDate(), charge, "Adjustment Item");
+						additionalItems.add(adjItem);
+						break;
+					}
+				}
+				break;
+			}
+		}
+
+		return additionalItems.build();
+	}
+
+	/**
+	 * This method returns all invoices of account
+	 *
+	 * @param account    The account to consider.
+	 * @param newInvoice New Invoice Item to be added to existing invoices of the
+	 *                   account
+	 * @param tenantCtx
+	 * @return All invoices of account
+	 */
+	private Set<Invoice> getAllInvoicesOfAccount(Account account, Invoice newInvoice, TenantContext tenantCtx) {
+		ImmutableSet.Builder<Invoice> builder = ImmutableSet.builder();
+		builder.addAll(getInvoicesByAccountId(account.getId(), tenantCtx));
+		builder.add(newInvoice);
+		return builder.build();
+	}
+
+	/**
+	 * Check whether adjustment item is already present in invoice Item of Invoice
+	 *
+	 * @param invoiceItems
+	 * @return
+	 */
+	private boolean checkforAdjustmentItem(List<InvoiceItem> invoiceItems) {
+		boolean adjustmentItemPresent = false;
+		for (InvoiceItem invoiceItem : invoiceItems) {
+			if (invoiceItem.getInvoiceItemType().equals(InvoiceItemType.ITEM_ADJ)) {
+				adjustmentItemPresent = true;
+			}
+		}
```

Comment:
Isn't there a bug here? If an `ITEM_ADJ` is found, shouldn't we break out of the for loop?

Reply:
Its not a bug. But yeah, you are right. We have put the break statement.